File: frontend/src/utils/searchTemplateTokens.ts
/**
 * Fallback catalog of every token a custom search query template understands.
 *
 * This mirrors SearchTemplateTokens.All in
 * src/Helpers/SearchTemplateTokens.cs, the backend's authoritative source of
 * truth. It exists as a *hardcoded* list - not derived from a successful
 * "GET /api/search/available-tokens" response - so the token picker still
 * shows every supported token when that request fails (offline, server
 * restarting, etc). Keep this list in sync with the backend catalog by hand;
 * SearchTemplateTokensTests on the backend and the tests alongside this file
 * both assert there are exactly 19 tokens.
 */
export interface SearchTemplateToken {
  token: string;
  description: string;
  example: string;
}

export const fallbackSearchTemplateTokens: SearchTemplateToken[] = [
  { token: '{League}', description: 'League name (normalized abbreviation)', example: 'NFL, UFC, Formula1' },
  { token: '{Year}', description: 'Event year (4 digits)', example: '2025' },
  { token: '{Month}', description: 'Event month (2 digits)', example: '01, 12' },
  { token: '{Day}', description: 'Event day (2 digits)', example: '01, 31' },
  { token: '{Round}', description: 'Round/race number, zero-padded (default, e.g. 01, 15)', example: '01, 15' },
  { token: '{Round:00}', description: 'Round/race number, zero-padded (same as {Round})', example: '01, 15' },
  { token: '{Round:0}', description: 'Round/race number, no padding', example: '1, 15' },
  { token: '{Week}', description: 'Week number (for team sports)', example: '1, 15' },
  { token: '{EventTitle}', description: 'Full event title (raw)', example: 'UFC 299, Super Bowl LVIII' },
  {
    token: '{EventName}',
    description:
      "Event title with trailing 'fighter1 vs fighter2' stripped (use for fighting cards where releases name the card, not the fighters)",
    example: "ONE Friday Fights 150 (from 'ONE Friday Fights 150 Kompetch vs Attachai')",
  },
  { token: '{Stage}', description: 'Stage number of a stage race, no padding; empty when the title names no stage', example: '16' },
  { token: '{Stage:00}', description: 'Stage number of a stage race, zero-padded', example: '16, 01' },
  { token: '{Stage:0}', description: 'Stage number of a stage race, no padding (same as {Stage})', example: '16' },
  { token: '{HomeTeam}', description: 'Home team name', example: 'Chiefs, Lakers' },
  { token: '{AwayTeam}', description: 'Away team name', example: 'Raiders, Celtics' },
  { token: '{vs}', description: 'Versus separator', example: 'vs' },
  { token: '{Season}', description: 'Season identifier', example: '2024-25, 2025' },
  { token: '{Part}', description: 'Part being searched (Prelims, Main Card, ...); empty for whole-event searches', example: 'Prelims, Main Card' },
  {
    token: '{EventType}',
    description: "Detected fighting event type, spaced for release-name matching; empty when the title doesn't classify",
    example: 'PPV, Fight Night, Contender Series, Weekly',
  },
];

File: src/Helpers/SearchTemplateTokens.cs
using Sportarr.Api.Models;

namespace Sportarr.Api.Helpers;

/// <summary>
/// The single authoritative catalog of every token a custom search query
/// template understands. Before this existed, three places each carried
/// their own idea of "the token list" - EventQueryService.BuildQueryFromTemplate
/// (19 tokens), the frontend token picker (16), and the
/// "GET /api/search/available-tokens" endpoint (12) - and they drifted out
/// of sync, so {Round:00}, {Stage:0}, and {vs} could be typed into a
/// template and honored by the builder but never offered by the UI.
///
/// This list is the one place that changes when a token is added, removed,
/// or documented differently. EventQueryService's replacement map and this
/// catalog's token set are asserted equal in
/// SearchTemplateTokensTests, so the two can never drift apart again.
/// </summary>
public static class SearchTemplateTokens
{
    public static readonly IReadOnlyList<SearchTemplateToken> All = new List<SearchTemplateToken>
    {
        new("{League}", "League name (normalized abbreviation)", "NFL, UFC, Formula1"),
        new("{Year}", "Event year (4 digits)", "2025"),
        new("{Month}", "Event month (2 digits)", "01, 12"),
        new("{Day}", "Event day (2 digits)", "01, 31"),
        new("{Round}", "Round/race number, zero-padded (default, e.g. 01, 15)", "01, 15"),
        new("{Round:00}", "Round/race number, zero-padded (same as {Round})", "01, 15"),
        new("{Round:0}", "Round/race number, no padding", "1, 15"),
        new("{Week}", "Week number (for team sports)", "1, 15"),
        new("{EventTitle}", "Full event title (raw)", "UFC 299, Super Bowl LVIII"),
        new("{EventName}", "Event title with trailing 'fighter1 vs fighter2' stripped (use for fighting cards where releases name the card, not the fighters)", "ONE Friday Fights 150 (from 'ONE Friday Fights 150 Kompetch vs Attachai')"),
        new("{Stage}", "Stage number of a stage race, no padding; empty when the title names no stage", "16"),
        new("{Stage:00}", "Stage number of a stage race, zero-padded", "16, 01"),
        new("{Stage:0}", "Stage number of a stage race, no padding (same as {Stage})", "16"),
        new("{HomeTeam}", "Home team name", "Chiefs, Lakers"),
        new("{AwayTeam}", "Away team name", "Raiders, Celtics"),
        new("{vs}", "Versus separator", "vs"),
        new("{Season}", "Season identifier", "2024-25, 2025"),
        new("{Part}", "Part being searched (Prelims, Main Card, ...); empty for whole-event searches", "Prelims, Main Card"),
        new("{EventType}", "Detected fighting event type, spaced for release-name matching; empty when the title doesn't classify", "PPV, Fight Night, Contender Series, Weekly"),
    };
}
